Keep track of a context for each listener

elupus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

* Coordinators that rely on internal _listeners and or overload the async_add_listener need to be adapted to take the additional context parameters
* Coordinators that override the async_remove_listener to detect when no listeners are active need to switch to using the _unschedule_refresh
* Coordinators that explicitly call async_remove_listener() should instead call the callback returned from async_add_listener()

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

* Allow a update coordinator to adapt what data to request on update from the backing service based on which entities are enabled.
* Remove async_remove_listener() and instead purely rely on return from async_add_listener(). This allow a better tracking sentinel for the registrations internally and allow for multiple registrations on same callback function.
* Add new overridable _unschedule_refresh() function to allow subclassing coordinators to know when there are no longer any listeners.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Updated developer docs: https://github.com/home-assistant/developers.home-assistant/pull/1355

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

elupus committed:
Keep track of a context for each listener

This allow a update coordinator to adapt what data to request on update from the backing service based on which entities are enabled.

elupus commented:
I would need some help on getting typehints correct for this. Preferable the type should only be set on the coordinator, then inherited by the CoordinatorEntity

balloob commented:
This is a nice addition 👍

elupus committed:
Clone list before calling callbacks

The callbacks can end up unregistering and modifying the dict while iterating.

elupus committed:
Only yield actual values

elupus committed:
Add a test for update context

elupus commented on homeassistant/helpers/update_coordinator.py:
```diff
@@ -82,11 +82,13 @@ def __init__(
         schedule_refresh = not self._listeners
 
-        self._listeners.append(update_callback)
+        self._listeners[update_callback] = context
```
This might be a bit problematic. The current code seem to handle passing the same update_callback many times, then i think you can call remove listener same amount of times to completely remove the the callback. It's now assumed to be unique.

elupus replied:
I've switched to using the "return cleanup" approach for the coordinator, that resolves this ambiguity and allow multiple calls on same update_callback (theoretically we could pass along the context, but that would be a rather large change for many components)

MartinHjelmare commented:
Just adding the label about docs so we don't forget to update the dev docs before we merge there. :+1:

elupus committed:
Remove async_remove_listener

This avoids the ambuigity as to what happens if same callback is added multiple times.

elupus commented:
I did realise that we could possibly achieve the same type of solution by annotating each update_callback function of an entity with some known attribute, similar to core.callback() uses setattr(func, "_hass_callback", True), but it would have to be instance (`self`) aware, so could get a bit ugly.

elupus committed:
Factor out iteration of _listeners to helper

balloob reviewed (approved):
Looks good! Can you also update the `test_coordinator_entity` test to make sure the context is passed in?

elupus commented:
Will do. Will look at getting a docs update too.

elupus committed:
Verify context is passed to coordinator

elupus committed:
Switch to Any as type instead of object

elupus committed:
Remove function which use was dropped earliers

The use was removed in 8bee25c938a123f0da7569b4e2753598d478b900
